parachain/integration-test/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+	let get_request = event_result_a
+		.into_iter()
+		.find_map(|event| match event {
+			Event::GetRequest(_) => Some(event),
+			_ => None,
+		})
+		.expect("Expected at least one GetRequest event");
+	// ======== process the request offchain ( 1. Make the response ) =============
+
+	let response = {
+		let get = match get_request {
+			Event::GetRequest(get) => get,
+			_ => panic!("Not supported"),
+		};
+		let dest_chain_block_hash = client_b.rpc().block_hash(Some(get.height.into())).await?;
+		let keys = get.keys.iter().map(|key| &key[..]).collect::<Vec<&[u8]>>();
+		let value_proof = client_b.rpc().read_proof(keys, dest_chain_block_hash).await?;
+		let proof = value_proof.proof.into_iter().map(|bytes| bytes.0).collect::<Vec<Vec<u8>>>();
+
+		let proof_of_value = SubstrateStateProof::StateProof(StateMachineProof {
+			hasher: HashAlgorithm::Keccak,
+			storage_proof: proof,
+		});
+		let proof = Proof {
+			height: StateMachineHeight {
+				id: chain_b_client.state_machine_id(),
+				height: get.height,
+			},
+			proof: proof_of_value.encode(),
+		};
+		let response = ResponseMessage {
+			datagram: RequestResponse::Request(vec![Request::Get(get.clone())]),
+			proof,
+			signer: chain_a_client.address(), // both A&B have same relayer address
+		};
+
+		Message::Response(response)
+	};
+	// =================== send to the source chain ================================
+	let _res = chain_a_client.submit(vec![response]).await?;
+	//==================== after approx 7-9 blocks the response event is emitted ===
+	// =================== fetch the returned value ================================
+
+	let mut response_event: Option<GetResponse> = None;
+
+	let mut state_machine_a_update_stream = chain_b_client
+		.state_machine_update_notification(chain_a_client.state_machine_id())
+		.await
+		.map_err(|err| anyhow!("StateMachineUpdated stream subscription failed: {err:?}"))?;
+
+	let mut block_height_correction = chain_a_client.query_finalized_height().await?;
+	while let Some(update_event_a) = state_machine_a_update_stream.next().await {
+		match update_event_a {
+			Ok(event_a) => {
+				// As it tracks finalized height some blocks will be skipped in the stream and they
+				// may contain the response
+				block_height_correction = block_height_correction
+					.checked_sub(event_a.latest_height)
+					.map_or(event_a.latest_height, |rem| {
+						if rem > 0 {
+							block_height_correction + 1
+						} else {
+							event_a.latest_height
+						}
+					});
+
+				let block_hash =
+					client_a.rpc().block_hash(Some(block_height_correction.into())).await?.unwrap();
+				let fetched_events =
+					client_a.events().at(block_hash).await?.find_first::<GetResponse>()?;
+
+				match fetched_events {
+					Some(res_event) => {
+						response_event = Some(res_event.clone());
+						break
+					},
+					None => continue,
+				}
+			},
+			Err(err) => {
+				panic!("Error in state machine update stream: {err:?}")
+			},
+		}
+	}
```

Comment:
This is a very convoluted way to achieve this.

Instead, subscribe to all new (finalized?) blocks and query events at those blocks.